File: plugins/shipyard/tests/test_auto_approve_data.py
#!/usr/bin/env python3
"""Tests for auto-approve-data.py PreToolUse hook."""

import json
import os
import subprocess
import sys
import tempfile
import unittest

SCRIPT = os.path.join(
    os.path.dirname(__file__),
    '..', 'project-files', 'scripts', 'auto-approve-data.py'
)


def run_hook(hook_input: dict, env_extra: dict | None = None) -> tuple[str, int]:
    """Run the hook script with given stdin JSON and env vars. Returns (stdout, exit_code)."""
    env = os.environ.copy()
    env.pop('SHIPYARD_DATA', None)
    if env_extra:
        env.update(env_extra)
    proc = subprocess.run(
        [sys.executable, SCRIPT],
        input=json.dumps(hook_input),
        capture_output=True,
        text=True,
        env=env,
    )
    return proc.stdout, proc.returncode


class TestAutoApproveData(unittest.TestCase):

    def setUp(self):
        # Real tempdir so realpath/commonpath have something to resolve
        self.tmpdir = tempfile.mkdtemp(prefix='shipyard-test-')
        self.sd = os.path.realpath(self.tmpdir)

    def tearDown(self):
        import shutil
        shutil.rmtree(self.tmpdir, ignore_errors=True)

    def test_approves_write_to_shipyard_data(self):
        """Edit to a file inside SHIPYARD_DATA should be auto-approved."""
        stdout, code = run_hook(
            {'tool_name': 'Edit', 'tool_input': {'file_path': os.path.join(self.sd, 'spec.md')}},
            {'SHIPYARD_DATA': self.sd},
        )
        self.assertEqual(code, 0)
        resp = json.loads(stdout)
        self.assertEqual(resp['hookSpecificOutput']['permissionDecision'], 'allow')

    def test_approves_write_to_nested_subdir(self):
        """Write to a deeply nested file inside SHIPYARD_DATA should be approved."""
        nested = os.path.join(self.sd, 'sprints', 's1', 'SPRINT.md')
        stdout, code = run_hook(
            {'tool_name': 'Write', 'tool_input': {'file_path': nested}},
            {'SHIPYARD_DATA': self.sd},
        )
        self.assertEqual(code, 0)
        resp = json.loads(stdout)
        self.assertEqual(resp['hookSpecificOutput']['permissionDecision'], 'allow')

    def test_rejects_write_outside_shipyard_data(self):
        """Edit to a file outside SHIPYARD_DATA should produce no output."""
        stdout, code = run_hook(
            {'tool_name': 'Edit', 'tool_input': {'file_path': '/home/user/project/src/main.py'}},
            {'SHIPYARD_DATA': self.sd},
        )
        self.assertEqual(code, 0)
        self.assertEqual(stdout, '')

    def test_rejects_path_traversal(self):
        """Path that escapes SHIPYARD_DATA via traversal should not be approved."""
        traversal = os.path.join(self.sd, '..', 'etc', 'passwd')
        stdout, code = run_hook(
            {'tool_name': 'Write', 'tool_input': {'file_path': traversal}},
            {'SHIPYARD_DATA': self.sd},
        )
        self.assertEqual(code, 0)
        self.assertEqual(stdout, '')

    def test_rejects_symlink_escape(self):
        """SECURITY: a symlink inside SHIPYARD_DATA pointing outside must NOT be approved.

        This is the CVE-2026-34452 class of bug — abspath() does not resolve
        symlinks, allowing prefix checks to be bypassed. realpath() must be used.
        """
        # Create a target outside SHIPYARD_DATA
        outside_dir = tempfile.mkdtemp(prefix='shipyard-outside-')
        try:
            # Plant a symlink inside SHIPYARD_DATA pointing to outside
            link_path = os.path.join(self.sd, 'evil')
            os.symlink(outside_dir, link_path)

            # Attempt to write to a file under the symlink (which resolves outside)
            target = os.path.join(link_path, 'pwned.txt')
            stdout, code = run_hook(
                {'tool_name': 'Write', 'tool_input': {'file_path': target}},
                {'SHIPYARD_DATA': self.sd},
            )
            self.assertEqual(code, 0)
            self.assertEqual(stdout, '', "Symlink escape should NOT be approved")
        finally:
            import shutil
            shutil.rmtree(outside_dir, ignore_errors=True)

    def test_ignores_non_file_tools(self):
        """Non-file tools (e.g., Bash) should be ignored silently."""
        stdout, code = run_hook(
            {'tool_name': 'Bash', 'tool_input': {'command': 'rm -rf /'}},
            {'SHIPYARD_DATA': self.sd},
        )
        self.assertEqual(code, 0)
        self.assertEqual(stdout, '')

    def test_ignores_missing_shipyard_data_env(self):
        """No SHIPYARD_DATA env var should produce no output."""
        stdout, code = run_hook(
            {'tool_name': 'Edit', 'tool_input': {'file_path': os.path.join(self.sd, 'spec.md')}},
        )
        self.assertEqual(code, 0)
        self.assertEqual(stdout, '')

    def test_ignores_missing_file_path(self):
        """Tool input without file_path should be ignored."""
        stdout, code = run_hook(
            {'tool_name': 'Edit', 'tool_input': {}},
            {'SHIPYARD_DATA': self.sd},
        )
        self.assertEqual(code, 0)
        self.assertEqual(stdout, '')

    def test_handles_invalid_json_input(self):
        """Invalid JSON on stdin should exit cleanly."""
        env = os.environ.copy()
        env.pop('SHIPYARD_DATA', None)
        proc = subprocess.run(
            [sys.executable, SCRIPT],
            input='not json',
            capture_output=True,
            text=True,
            env=env,
        )
        self.assertEqual(proc.returncode, 0)
        self.assertEqual(proc.stdout, '')

    def test_read_tool_no_longer_approved(self):
        """SECURITY: Read is not in the auto-approve matcher anymore (scope reduction)."""
        stdout, code = run_hook(
            {'tool_name': 'Read', 'tool_input': {'file_path': os.path.join(self.sd, 'backlog.md')}},
            {'SHIPYARD_DATA': self.sd},
        )
        self.assertEqual(code, 0)
        self.assertEqual(stdout, '', "Read should NOT be auto-approved (removed from matcher)")

    def test_prefix_attack_not_approved(self):
        """Path with SHIPYARD_DATA as string prefix but in a sibling dir should not be approved."""
        # Create a sibling directory with a similar name
        parent = os.path.dirname(self.sd)
        sibling = self.sd + '-evil'
        os.makedirs(sibling, exist_ok=True)
        try:
            stdout, code = run_hook(
                {'tool_name': 'Write', 'tool_input': {'file_path': os.path.join(sibling, 'hack.py')}},
                {'SHIPYARD_DATA': self.sd},
            )
            self.assertEqual(code, 0)
            self.assertEqual(stdout, '')
        finally:
            import shutil
            shutil.rmtree(sibling, ignore_errors=True)

    def test_dotdot_segment_rejected_pre_resolution(self):
        """Defense in depth: '..' segments rejected before path resolution."""
        target = self.sd + '/subdir/../../../etc/passwd'
        stdout, code = run_hook(
            {'tool_name': 'Write', 'tool_input': {'file_path': target}},
            {'SHIPYARD_DATA': self.sd},
        )
        self.assertEqual(code, 0)
        self.assertEqual(stdout, '')


if __name__ == '__main__':
    unittest.main()

File: plugins/shipyard/tests/test_auto_approve_data.mjs
/**
 * Parity tests for bin/hooks/auto-approve-data.mjs.
 *
 * Mirrors tests/test_auto_approve_data.py case-for-case. Run via:
 *   node --test tests/test_auto_approve_data.mjs
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, rmSync, mkdirSync, symlinkSync, existsSync, readFileSync, realpathSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { run as runHook } from "../bin/hooks/auto-approve-data.mjs";

async function withTempDir(fn) {
  const dir = mkdtempSync(join(tmpdir(), "auto-approve-test-"));
  const real = realpathSync(dir);
  try {
    return await fn(real);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
}

async function runWithEnv(hookInput, env) {
  // Capture stdout from the hook by stubbing process.stdout.write.
  const originalWrite = process.stdout.write.bind(process.stdout);
  const captured = [];
  process.stdout.write = (chunk) => {
    captured.push(typeof chunk === "string" ? chunk : chunk.toString("utf8"));
    return true;
  };
  // Set env vars for the duration of the call.
  const orig = {};
  for (const k of Object.keys(env || {})) {
    orig[k] = process.env[k];
    process.env[k] = env[k];
  }
  // Always clear SHIPYARD_DATA unless overridden, to match Python test setup.
  if (!("SHIPYARD_DATA" in (env || {}))) {
    orig.SHIPYARD_DATA = process.env.SHIPYARD_DATA;
    delete process.env.SHIPYARD_DATA;
  }
  let code;
  try {
    code = await runHook(hookInput, process.env);
  } finally {
    process.stdout.write = originalWrite;
    for (const k of Object.keys(orig)) {
      if (orig[k] === undefined) delete process.env[k];
      else process.env[k] = orig[k];
    }
  }
  return { stdout: captured.join(""), code };
}

test("auto-approve: approves Edit to file inside SHIPYARD_DATA", async () => {
  await withTempDir(async (sd) => {
    const { stdout, code } = await runWithEnv(
      { tool_name: "Edit", tool_input: { file_path: join(sd, "spec.md") } },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    const resp = JSON.parse(stdout);
    assert.equal(resp.hookSpecificOutput.permissionDecision, "allow");
  });
});

test("auto-approve: approves Write to nested subdir", async () => {
  await withTempDir(async (sd) => {
    const nested = join(sd, "sprints", "s1", "SPRINT.md");
    const { stdout, code } = await runWithEnv(
      { tool_name: "Write", tool_input: { file_path: nested } },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    const resp = JSON.parse(stdout);
    assert.equal(resp.hookSpecificOutput.permissionDecision, "allow");
  });
});

test("auto-approve: approves MultiEdit (not just Edit) to SHIPYARD_DATA", async () => {
  await withTempDir(async (sd) => {
    const { stdout, code } = await runWithEnv(
      { tool_name: "MultiEdit", tool_input: { file_path: join(sd, "spec.md") } },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    const resp = JSON.parse(stdout);
    assert.equal(resp.hookSpecificOutput.permissionDecision, "allow");
  });
});

test("auto-approve: rejects write outside SHIPYARD_DATA (no JSON output)", async () => {
  await withTempDir(async (sd) => {
    const { stdout, code } = await runWithEnv(
      {
        tool_name: "Edit",
        tool_input: { file_path: "/home/user/project/src/main.py" },
      },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    assert.equal(stdout, "");
  });
});

test("auto-approve: rejects path with .. traversal segment", async () => {
  await withTempDir(async (sd) => {
    const traversal = join(sd, "..", "etc", "passwd");
    const { stdout, code } = await runWithEnv(
      { tool_name: "Write", tool_input: { file_path: traversal } },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    assert.equal(stdout, "");
  });
});

test("auto-approve: rejects symlink escape (TOCTOU defense)", async () => {
  await withTempDir(async (sd) => {
    const outside = mkdtempSync(join(tmpdir(), "outside-"));
    try {
      const link = join(sd, "evil");
      symlinkSync(outside, link);
      const target = join(link, "pwned.txt");
      const { stdout, code } = await runWithEnv(
        { tool_name: "Write", tool_input: { file_path: target } },
        { SHIPYARD_DATA: sd },
      );
      assert.equal(code, 0);
      assert.equal(stdout, "", "symlink escape must NOT be approved");
    } finally {
      rmSync(outside, { recursive: true, force: true });
    }
  });
});

test("auto-approve: ignores non-file tools (Bash)", async () => {
  await withTempDir(async (sd) => {
    const { stdout, code } = await runWithEnv(
      { tool_name: "Bash", tool_input: { command: "rm -rf /" } },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    assert.equal(stdout, "");
  });
});

test("auto-approve: ignores missing file_path", async () => {
  await withTempDir(async (sd) => {
    const { stdout, code } = await runWithEnv(
      { tool_name: "Edit", tool_input: {} },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    assert.equal(stdout, "");
  });
});

test("auto-approve: Read tool no longer approved (scope reduction)", async () => {
  await withTempDir(async (sd) => {
    const { stdout, code } = await runWithEnv(
      {
        tool_name: "Read",
        tool_input: { file_path: join(sd, "backlog.md") },
      },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    assert.equal(stdout, "");
  });
});

test("auto-approve: prefix attack (sibling -evil dir) not approved", async () => {
  await withTempDir(async (sd) => {
    const sibling = sd + "-evil";
    mkdirSync(sibling, { recursive: true });
    try {
      const { stdout, code } = await runWithEnv(
        {
          tool_name: "Write",
          tool_input: { file_path: join(sibling, "hack.py") },
        },
        { SHIPYARD_DATA: sd },
      );
      assert.equal(code, 0);
      assert.equal(stdout, "");
    } finally {
      rmSync(sibling, { recursive: true, force: true });
    }
  });
});

test("auto-approve: dotdot segments rejected pre-resolution", async () => {
  await withTempDir(async (sd) => {
    const target = sd + "/subdir/../../../etc/passwd";
    const { stdout, code } = await runWithEnv(
      { tool_name: "Write", tool_input: { file_path: target } },
      { SHIPYARD_DATA: sd },
    );
    assert.equal(code, 0);
    assert.equal(stdout, "");
  });
});

test("auto-approve: breadcrumb log written on allow", async () => {
  await withTempDir(async (sd) => {
    await runWithEnv(
      {
        tool_name: "Edit",
        tool_input: { file_path: join(sd, "spec.md") },
      },
      { SHIPYARD_DATA: sd },
    );
    const log = join(sd, ".auto-approve.log");
    assert.ok(existsSync(log), "breadcrumb log should be created");
    const content = readFileSync(log, "utf8");
    assert.ok(content.includes("allow"));
    assert.ok(content.includes("Edit"));
  });
});

test("auto-approve: breadcrumb creates data dir if missing (R12)", async () => {
  await withTempDir(async (sd) => {
    const fresh = join(sd, "fresh-data-dir");
    assert.equal(existsSync(fresh), false);
    const target = join(fresh, "spec.md");
    await runWithEnv(
      { tool_name: "Edit", tool_input: { file_path: target } },
      { SHIPYARD_DATA: fresh },
    );
    const log = join(fresh, ".auto-approve.log");
    assert.ok(
      existsSync(log),
      "first-run breadcrumb should create the data dir and write the log",
    );
  });
});

test("auto-approve: breadcrumb log written on pass (file outside data dir)", async () => {
  await withTempDir(async (sd) => {
    const outside = mkdtempSync(join(tmpdir(), "outside-pass-"));
    try {
      await runWithEnv(
        {
          tool_name: "Edit",
          tool_input: { file_path: join(outside, "main.py") },
        },
        { SHIPYARD_DATA: sd },
      );
      const log = join(sd, ".auto-approve.log");
      assert.ok(existsSync(log));
      const content = readFileSync(log, "utf8");
      assert.ok(content.includes("pass"));
    } finally {
      rmSync(outside, { recursive: true, force: true });
    }
  });
});

// --- CLI-owned state deny (v2.9.0) ------------------------------------
//
// Cursor files, PROGRESS.md and HANDOFF.md have a single deterministic
// writer: the shipyard-data CLI (which runs the terminal-evidence gate and
// loop-leak guard in-process — see test_cursor_cli.mjs). The hook's job is
// now simply to DENY any model Write/Edit to these basenames inside the
// data dir, with a hint pointing at the CLI. These tests replace the
// v2.6.0 gate-wiring and v2.8.2 loop-leak hook tests — that logic moved
// into `shipyard-data cursor advance`.

import { mkdirSync as mkdirSyncFs } from "node:fs";

const CLI_OWNED = ["EXECUTE-CURSOR.md", "REVIEW-CURSOR.md", "PROGRESS.md", "HANDOFF.md"];

for (const base of CLI_OWNED) {
  test(`cli-owned deny: Write to ${base} in the data dir is DENIED with a CLI hint`, async () => {
    await withTempDir(async (sd) => {
      mkdirSyncFs(join(sd, "sprints", "current"), { recursive: true });
      const { stdout, code } = await runWithEnv(
        {
          tool_name: "Write",
          tool_input: {
            file_path: join(sd, "sprints", "current", base),
            content: "anything",
          },
        },
        { SHIPYARD_DATA: sd },
      );
      assert.equal(code, 0);
      const resp = JSON.parse(stdout);
      assert.equal(resp.hookSpecificOutput.permissionDecision, "deny");
      assert.match(
        resp.hookSpecificOutput.permissionDecisionReason,
        /shipyard-data cursor/,
        "deny reason must point the model at the CLI",
      );
    });
  });
}

test("cli-owned deny: Edit to a cursor file is DENIED (not just Write)", async () => {
  await withTempDir(async (sd) => {
    mkdirSyncFs(join(sd, "sprints", "current"), { recursive: true });
    const { stdout } = await runWithEnv(
      {
        tool_name: "Edit",
        tool_input: {
          file_path: join(sd, "sprints", "current", "EXECUTE-CURSOR.md"),
          old_string: "terminal: false",
          new_string: "terminal: true",
        },
      },
      { SHIPYARD_DATA: sd },
    );
    const resp = JSON.parse(stdout);
    assert.equal(resp.hookSpecificOutput.permissionDecision, "deny");
  });
});

test("cli-owned deny: same basename OUTSIDE the data dir passes through (no deny)", async () => {
  // A user project can legitimately have its own PROGRESS.md — the deny is
  // scoped to the Shipyard data dir.
  await withTempDir(async (sd) => {
    const outside = mkdtempSync(join(tmpdir(), "outside-progress-"));
    try {
      const { stdout, code } = await runWithEnv(
        {
          tool_name: "Write",
          tool_input: { file_path: join(outside, "PROGRESS.md"), content: "x" },
        },
        { SHIPYARD_DATA: sd },
      );
      assert.equal(code, 0);
      assert.equal(stdout, "", "outside the data dir the hook stays silent (default permission flow)");
    } finally {
      rmSync(outside, { recursive: true, force: true });
    }
  });
});

test("cli-owned deny: other data-dir files (e.g. SPRINT.md body, spec files) still auto-approved", async () => {
  await withTempDir(async (sd) => {
    mkdirSyncFs(join(sd, "sprints", "current"), { recursive: true });
    const { stdout } = await runWithEnv(
      {
        tool_name: "Write",
        tool_input: {
          file_path: join(sd, "sprints", "current", "SPRINT.md"),
          content: "---\nid: sprint-001\n---\n\n### Wave 1\nTasks: [T001]\n",
        },
      },
      { SHIPYARD_DATA: sd },
    );
    const resp = JSON.parse(stdout);
    assert.equal(resp.hookSpecificOutput.permissionDecision, "allow");
  });
});

test("cli-owned deny: breadcrumb log records the deny", async () => {
  await withTempDir(async (sd) => {
    mkdirSyncFs(join(sd, "sprints", "current"), { recursive: true });
    await runWithEnv(
      {
        tool_name: "Write",
        tool_input: {
          file_path: join(sd, "sprints", "current", "REVIEW-CURSOR.md"),
          content: "x",
        },
      },
      { SHIPYARD_DATA: sd },
    );
    const log = readFileSync(join(sd, ".auto-approve.log"), "utf8");
    assert.ok(log.includes("cli_owned_state"));
  });
});
